### PR TITLE
Add logger-handler to firewall

### PR DESCRIPTION
**Pull request by storojs72**

### CODE REVIEW

File: cmd/acraserver/config.go [hunk 1]
```diff
@@ -70,8 +71,26 @@ func NewConfig() *Config {
 
 var ErrTwoDBSetup = errors.New("only one db supported at one time")
 
-func (config *Config) SetFirewall(fw firewall.FirewallInterface) {
-	config.firewall = fw
+func (config *Config) SetFirewall(censorConfigPath string) error {
+	firewall := &firewall.Firewall{}
```
Comment (Lagovas) [inline cmd/acraserver/config.go]: better to assign one time `config.firewall = &firewall.Firewall{}` instead 5 extra lines

File: cmd/acraserver/acraserver.go [hunk 2]
```diff
@@ -148,6 +149,12 @@ func main() {
 		os.Exit(1)
 	}
 
+	if err := config.SetFirewall(*firewallConfig); err != nil {
```
Comment (vixentael) [inline cmd/acraserver/acraserver.go]: I think we shouldn't return error if no config file.
users should be able not to start firewall if they don't want to.

i would check if `firewallConfig` exists first, then `setFirewall`

File: firewall/handlers/logging_handler.go [hunk 3]
```diff
@@ -0,0 +1,89 @@
...
+	return queries
+}
+
+func (handler *LoggingHandler) MarkQueryAsForbidden(query string) {
+	for index, queryInfo := range handler.Queries {
+		if strings.EqualFold(query, queryInfo.RawQuery) {
+			handler.Queries[index].IsForbidden = true
+		}
+	}
+	handler.Serialize()
```
Comment (Lagovas) [inline firewall/handlers/logging_handler.go]: what if error?

File: firewall/handlers/logging_handler.go [hunk 4]
```diff
@@ -0,0 +1,89 @@
...
+		if queryInfo.IsForbidden == true{
+			forbiddenQueries = append(forbiddenQueries, queryInfo.RawQuery)
+		}
+	}
+	return forbiddenQueries
+}
+
+func (handler *LoggingHandler) Serialize() error {
+	jsonFile, err := json.Marshal(handler.Queries)
+	err = ioutil.WriteFile(handler.filePath, jsonFile, 0600)
```
Comment (Lagovas) [inline firewall/handlers/logging_handler.go]: you can return result because it err itself

File: firewall/handlers/logging_handler.go [hunk 5]
```diff
@@ -0,0 +1,89 @@
...
+	return nil
+}
+
+func (handler *LoggingHandler) Deserialize() error {
+	var bufferBytes []byte
+	bufferBytes, err := ioutil.ReadFile(handler.filePath)
+	if err != nil {
+		return err
+	}
+	json.Unmarshal(bufferBytes, &handler.Queries)
```
Comment (Lagovas) [inline firewall/handlers/logging_handler.go]: `err := json.Unmarshal(...)`

File: firewall/handlers/logging_handler.go [hunk 6]
```diff
@@ -0,0 +1,89 @@
...
+	if err != nil {
```
Comment (Lagovas) [inline firewall/handlers/logging_handler.go]: here you check ReadFile call

**Comment by vixentael** [comment]
Please add info log to see if query was allowed/filtered here:
https://github.com/cossacklabs/acra/blob/master/firewall/firewall_implementation.go#L23

And show example of json output

File: acracensor/acracensor_implementation.go [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+func (acraCensor *AcraCensor) RemoveHandler(handler QueryHandlerInterface) {
+	for index, handlerFromRange := range acraCensor.handlers {
+		if handlerFromRange == handler {
+			acraCensor.handlers = append(acraCensor.handlers[:index], acraCensor.handlers[index+1:]...)
+		}
+	}
+}
+
+func (acraCensor *AcraCensor) HandleQuery(query string) error {
+	log.Infof("Censor works")
```
Comment (vixentael) [inline acracensor/acracensor_implementation.go]: we will receive this log message on every query. which is not very useful, i'd remove it

File: acracensor/acracensor_implementation.go [hunk 8]
```diff
@@ -0,0 +1,31 @@
...
+	for _, handler := range acraCensor.handlers {
+		log.Infof("Handler: %s", handler.GetName())
+		if err := handler.CheckQuery(query); err != nil {
+			return err
```
Comment (vixentael) [inline acracensor/acracensor_implementation.go]: `log.WithFields(log.Fields{"acracensor" : "forbidden query", "handler" : handler.GetName()}).Infof("Query: %s", query)`

File: acracensor/acracensor_implementation.go [hunk 9]
```diff
@@ -0,0 +1,31 @@
...
+		if handlerFromRange == handler {
+			acraCensor.handlers = append(acraCensor.handlers[:index], acraCensor.handlers[index+1:]...)
+		}
+	}
+}
+
+func (acraCensor *AcraCensor) HandleQuery(query string) error {
+	log.Infof("Censor works")
+	for _, handler := range acraCensor.handlers {
+		log.Infof("Handler: %s", handler.GetName())
```
Comment (vixentael) [inline acracensor/acracensor_implementation.go]: i think we should remove this

File: acracensor/acracensor_implementation.go [hunk 10]
```diff
@@ -0,0 +1,31 @@
...
+		if err := handler.CheckQuery(query); err != nil {
+			return err
+		}
```
Comment (vixentael) [inline acracensor/acracensor_implementation.go]: `log.WithFields(log.Fields{"acracensor" : handler.GetName()}).Infof("Allowed query: %s", query)`

File: cmd/acraserver/config.go [hunk 11]
```diff
@@ -50,7 +51,7 @@ type Config struct {
 	postgresql              bool
 	configPath              string
 	debug                   bool
-	firewall                firewall.FirewallInterface
+	firewall                acracensor.AcracensorInterface
```
Comment (vixentael) [inline cmd/acraserver/config.go]: `censor` instead of `firewall`?

File: logging/event_codes.go [hunk 12]
```diff
@@ -52,8 +52,9 @@ const (
 	EventCodeErrorCantParseAuthData       = 557
 	EventCodeErrorCantDumpConfig          = 558
 
-	// firewall
+	// acracensor
 	EventCodeErrorFirewallQueryIsNotAllowed = 560
```
Comment (vixentael) [inline logging/event_codes.go]: censor instead of firewall?

Comment (storojs72) [reply logging/event_codes.go]: Agree

File: cmd/acraserver/config.go (hunk 11 above)
Comment (storojs72) [reply cmd/acraserver/config.go]: Agree

File: cmd/acraserver/config.go [hunk 13]
```diff
@@ -70,11 +71,26 @@ func NewConfig() *Config {
...
+	}
+	configuration, err := ioutil.ReadFile(censorConfigPath)
+	if err != nil {
+		return err
+	}
+	err = acraCensor.LoadConfiguration(configuration)
+	if err != nil {
+		return err
+	}
+	config.censor = acraCensor
```
Comment (Lagovas) [inline cmd/acraserver/config.go]: unnecessary

File: acracensor/acracensor_implementation.go [hunk 14]
```diff
@@ -0,0 +1,30 @@
...
+		if handlerFromRange == handler {
+			acraCensor.handlers = append(acraCensor.handlers[:index], acraCensor.handlers[index+1:]...)
+		}
+	}
+}
+
+func (acraCensor *AcraCensor) HandleQuery(query string) error {
+	for _, handler := range acraCensor.handlers {
+		if err := handler.CheckQuery(query); err != nil {
+			return err
```
Comment (Lagovas) [inline acracensor/acracensor_implementation.go]: log error. will be good to log handler name too

**Pull request merged**